### ISSUE TITLE
React proxy errors even though flask is running

### ISSUE THREAD

**Issue by SashvDave** [princeton-nlp/SWE-agent#449]
### Describe the bug

Error Log:
"Failed to connect to the flask server, please restart it.
If you think this was a bug, please head over to [our GitHub issue tracker](https://github.com/princeton-nlp/swe-agent/issues), check if someone has already reported the issue, and if not, create a new issue. Please include the full log, all settings that you entered, and a screenshot of this page."

<img width="1149" alt="Screenshot 2024-05-28 at 6 15 18 PM" src="https://github.com/princeton-nlp/SWE-agent/assets/55955567/1bb836dc-b5ff-4319-85fc-f87c00445f3f">

### Steps/commands/code to Reproduce

```
docker run -p 3000:3000 -it -v /var/run/docker.sock:/var/run/docker.sock \
  -e OPENAI_API_KEY="sk-xxxxx" \
  sweagent/swe-agent-run:latest bash start_web_ui.sh \  
  python run.py --image_name=sweagent/swe-agent:latest \
  --model_name gpt4 \
  --data_path https://github.com/pvlib/pvlib-python/issues/1603 \
  --config_file config/default_from_url.yaml  --skip_existing=False
```

### Error message/results

```
Failed to connect to the flask server, please restart it.
If you think this was a bug, please head over to [our GitHub issue tracker](https://github.com/princeton-nlp/swe-agent/issues), check if someone has already reported the issue, and if not, create a new issue. Please include the full log, all settings that you entered, and a screenshot of this page.
```

### System Information

MacOS M1

### Checklist

- [X] I'm running with the latest docker container/on the latest development version
- [X] I've searched the other issues for a duplicate
- [X] I have copied the full command/code that I ran (as text, not as screenshot!)
- [X] If applicable: I have copied the **full** log file/error message that was the result (as text, not as screenshot!)
- [X] I have enclosed code/log messages in triple backticks ([docs](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code)) and clicked "Preview" to make sure it's displayed correctly.

**Comment by klieret** [comment]
Hi @SashvDave, thanks for the feedback. 

First, you only need 

```bash
docker run -p 3000:3000 -it -v /var/run/docker.sock:/var/run/docker.sock \
  -e OPENAI_API_KEY="sk-xxxxx" \
  sweagent/swe-agent-run:latest bash start_web_ui.sh
```

(not the `python run.py` etc.).

Second, what was the output of the docker command? Unfortunately I don't think I can help without it. It should have information about why the backend failed.

**Comment by SashvDave** [comment]
Hey @klieret, thank you for the help!

The following is what I receive in my terminal when I run the docker command:
```
(env) sashvdave@Sashvs-MacBook-Pro SWE-agent % docker run -p 3000:3000 -it -v /var/run/docker.sock:/var/run/docker.sock \
   -e OPENAI_API_KEY="sk-xxx" \
  sweagent/swe-agent-run:latest bash start_web_ui.sh

up to date, audited 1819 packages in 4s

349 packages are looking for funding
  run `npm fund` for details

8 vulnerabilities (2 moderate, 6 high)

To address all issues (including breaking changes), run:
  npm audit fix --force

Run `npm audit` for details.

                        -------------

__/\\\\\\\\\\\\\____/\\\\____________/\\\\____/\\\\\\\\\_____
 _\/\\\/////////\\\_\/\\\\\\________/\\\\\\__/\\\///////\\\___
  _\/\\\_______\/\\\_\/\\\//\\\____/\\\//\\\_\///______\//\\\__
   _\/\\\\\\\\\\\\\/__\/\\\\///\\\/\\\/_\/\\\___________/\\\/___
    _\/\\\/////////____\/\\\__\///\\\/___\/\\\________/\\\//_____
     _\/\\\_____________\/\\\____\///_____\/\\\_____/\\\//________
      _\/\\\_____________\/\\\_____________\/\\\___/\\\/___________
       _\/\\\_____________\/\\\_____________\/\\\__/\\\\\\\\\\\\\\\_
        _\///______________\///______________\///__\///////////////__


                          Runtime Edition

        PM2 is a Production Process Manager for Node.js applications
                     with a built-in Load Balancer.

                Start and Daemonize any application:
                $ pm2 start app.js

                Load Balance 4 instances of api.js:
                $ pm2 start api.js -i 4

                Monitor in production:
                $ pm2 monitor

                Make pm2 auto-boot at server restart:
                $ pm2 startup

                To go further checkout:
                http://pm2.io/


                        -------------

[PM2] Spawning PM2 daemon with pm2_home=/root/.pm2
[PM2] PM2 Successfully daemonized
[PM2] Starting /usr/bin/npm in fork_mode (1 instance)
[PM2] Done.
┌────┬────────────────────┬──────────┬──────┬───────────┬──────────┬──────────┐
│ id │ name               │ mode     │ ↺    │ status    │ cpu      │ memory   │
├────┼────────────────────┼──────────┼──────┼───────────┼──────────┼──────────┤
│ 0  │ swe-agent          │ fork     │ 0    │ online    │ 0%       │ 23.4mb   │
└────┴────────────────────┴──────────┴──────┴───────────┴──────────┴──────────┘
* If you are running on your own machine, then a browser window 
  should have already opened. If not, wait a few more seconds, then 
  open your browser at http://localhost:3000
* If you are running in github codespaces, please click the popup 
  that offers to forward port 3000 (not 8000!).
  Missed it? Find more information at 
  https://princeton-nlp.github.io/SWE-agent/installation/codespaces#running-the-web-ui
* Something went wrong? Please check 
  web_api.log for error messages!

```

I get no error messages in my web_api.log

**Comment by klieret** [comment]
Hmmm. Perhaps it's actually running but the web interface is confused... I've had this a few times and just had to wait for a few seconds... But perhaps I should revisit this and make it more stable. 

The best way is to set up swe-agent [from source](https://princeton-nlp.github.io/SWE-agent/installation/source/) and then manually start both parts of the web server ([see here](https://princeton-nlp.github.io/SWE-agent/usage/web_ui/))

**Comment by SashvDave** [comment]
Hmm, I tried starting both parts of the web server individually, still getting the same issue - i wait for a few seconds and it still doesn't go away. Here's the output I receive in my terminal:
```
Proxy error: Could not proxy request /socket.io/?EIO=4&transport=polling&t=O_88m_u from localhost:3000 to http://localhost:8000.
See https://nodejs.org/api/errors.html#errors_common_system_errors for more information (ECONNREFUSED).
```

I checked out the link but it just provides the error code, not the solution. tried a few solutions from stack overflow but no bueno

**Comment by klieret** [comment]
Were you starting them from a development install or also with the docker container?

Also, was there any more output? 

For example, if your port 8000 is taken, the `flask` backend will serve somewhere else and the frontend doesn't know about this...

**Comment by klieret** [comment]
Let me perhaps add a default route to the backend so you can check in your web server if it running.

**Comment by SashvDave** [comment]
Hey, sorry I missed your earlier message.

Yes, I am starting from a development install, not with the docker container.

No more output, that was it.

**Comment by SashvDave** [comment]
Hmm, I may have found the issue:
It doesn't log any issues when I open the server, but I navigated to the URL and it seems to have an issue. Pasted below
```
2024-06-01 11:38:32,893 - werkzeug - INFO - WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
 * Running on http://127.0.0.1:8000
2024-06-01 11:38:32,893 - werkzeug - INFO - Press CTRL+C to quit
2024-06-01 11:38:32,894 - werkzeug - INFO -  * Restarting with stat
2024-06-01 11:38:34,529 - werkzeug - WARNING -  * Debugger is active!
2024-06-01 11:38:34,540 - werkzeug - INFO -  * Debugger PIN: 128-207-556
2024-06-01 11:38:41,670 - werkzeug - INFO - 127.0.0.1 - - [01/Jun/2024 11:38:41] "GET / HTTP/1.1" 500 -
Traceback (most recent call last):
  File "/Users/sashvdave/SWE-agent/env/lib/python3.12/site-packages/flask/app.py", line 1498, in __call__
    return self.wsgi_app(environ, start_response)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/sashvdave/SWE-agent/env/lib/python3.12/site-packages/flask_socketio/__init__.py", line 43, in __call__
    return super(_SocketIOMiddleware, self).__call__(environ,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/sashvdave/SWE-agent/env/lib/python3.12/site-packages/engineio/middleware.py", line 74, in __call__
    return self.wsgi_app(environ, start_response)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/sashvdave/SWE-agent/env/lib/python3.12/site-packages/flask/app.py", line 1476, in wsgi_app
    response = self.handle_exception(e)
               ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/sashvdave/SWE-agent/env/lib/python3.12/site-packages/flask_cors/extension.py", line 178, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
                                                ^^^^^^^^^^^^^^^^^^^^
  File "/Users/sashvdave/SWE-agent/env/lib/python3.12/site-packages/flask/app.py", line 1473, in wsgi_app
    response = self.full_dispatch_request()
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/sashvdave/SWE-agent/env/lib/python3.12/site-packages/flask/app.py", line 882, in full_dispatch_request
    rv = self.handle_user_exception(e)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/sashvdave/SWE-agent/env/lib/python3.12/site-packages/flask_cors/extension.py", line 178, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
                                                ^^^^^^^^^^^^^^^^^^^^
  File "/Users/sashvdave/SWE-agent/env/lib/python3.12/site-packages/flask/app.py", line 880, in full_dispatch_request
    rv = self.dispatch_request()
         ^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/sashvdave/SWE-agent/env/lib/python3.12/site-packages/flask/app.py", line 865, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/sashvdave/SWE-agent/sweagent/api/server.py", line 102, in index
    return render_template("index.html")
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/sashvdave/SWE-agent/env/lib/python3.12/site-packages/flask/templating.py", line 149, in render_template
    template = app.jinja_env.get_or_select_template(template_name_or_list)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/sashvdave/SWE-agent/env/lib/python3.12/site-packages/jinja2/environment.py", line 1084, in get_or_select_template
    return self.get_template(template_name_or_list, parent, globals)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/sashvdave/SWE-agent/env/lib/python3.12/site-packages/jinja2/environment.py", line 1013, in get_template
    return self._load_template(name, globals)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/sashvdave/SWE-agent/env/lib/python3.12/site-packages/jinja2/environment.py", line 972, in _load_template
    template = self.loader.load(self, name, self.make_globals(globals))
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/sashvdave/SWE-agent/env/lib/python3.12/site-packages/jinja2/loaders.py", line 126, in load
    source, filename, uptodate = self.get_source(environment, name)
                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/sashvdave/SWE-agent/env/lib/python3.12/site-packages/flask/templating.py", line 65, in get_source
    return self._get_source_fast(environment, template)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/sashvdave/SWE-agent/env/lib/python3.12/site-packages/flask/templating.py", line 99, in _get_source_fast
    raise TemplateNotFound(template)
jinja2.exceptions.TemplateNotFound: index.html
```

**Comment by kwight** [comment]
I've also been getting tripped up by Flask server startup and existing processes when using `start_web_ui.sh`; this might help avoid some confusion: https://github.com/princeton-nlp/SWE-agent/pull/479

**Comment by klieret** [comment]
>  It doesn't log any issues when I open the server, but I navigated to the URL and it seems to have an issue. Pasted below

Yes, that's to be expected, currently we only serve the API on port 8000, not a website. But I've opened https://github.com/princeton-nlp/SWE-agent/issues/490 to do that for debugging purposes (trying to get this done today or tmr.)

Just to double check, you are opening your browser at `http://localhost:3000/`, right? (not 8000).

While doing that, could you check that the backend is running by checking

```bash
lsof -i :8000
```

(this checks for anything serving port 8000. it should show the flask server).

**Comment by ilansh** [comment]
Hey, I'm facing the same issue.  
Also tried running the service and frontend individually outside of docker, no luck.

**Comment by klieret** [comment]
Strange, strange. Let me make it a priority to get to the bottom of this.

Possible solutions:

* https://stackoverflow.com/a/63231052/ (use 127...)
* https://stackoverflow.com/a/57532447/ (trailing slash)
* https://stackoverflow.com/a/46574808/ (`--ignore-client`)
* https://stackoverflow.com/a/73739848/ (change axio configuration)
* https://create-react-app.dev/docs/proxying-api-requests-in-development/#configuring-the-proxy-manually

Let me attempt the first two for now

**Comment by ilansh** [comment]
yup I just solved it by using http://127.0.0.1:8000 as the proxy in `package.json`.
Thanks!
